backend/inspirehep/alembic/6063cc09aec9_add_students_advisors_table.py

```diff
@@ -0,0 +1,70 @@
+#
+# This file is part of Invenio.
+# Copyright (C) 2016-2018 CERN.
+#
+# Invenio is free software; you can redistribute it and/or modify it
+# under the terms of the MIT License; see LICENSE file for more details.
+
+"""add students_advisors table"""
+
+import sqlalchemy as sa
+from alembic import op
+from sqlalchemy.dialects import postgresql
+from sqlalchemy_utils.types import UUIDType
+
+# revision identifiers, used by Alembic.
+revision = "6063cc09aec9"
+down_revision = "2d7ea622feda"
+branch_labels = ()
+depends_on = None
+
+
+def upgrade():
+    """Upgrade database."""
+    # ### commands auto generated by Alembic - please adjust! ###
+    op.create_table(
+        "students_advisors",
+        sa.Column("advisor_id", UUIDType, nullable=False),
+        sa.Column("student_id", UUIDType, nullable=False),
+        sa.Column(
+            "degree_type",
+            postgresql.ENUM(
+                "other",
+                "diploma",
+                "bachelor",
+                "laurea",
+                "master",
+                "phd",
+                "habilitation",
+                name="enum_degree_type",
+            ),
+            nullable=False,
+        ),
+        sa.ForeignKeyConstraint(
+            ["advisor_id"],
+            ["records_metadata.id"],
+            name="fk_students_advisors_advisor_id",
+        ),
+        sa.ForeignKeyConstraint(
+            ["student_id"],
+            ["records_metadata.id"],
+            name="fk_students_advisors_student_id",
+        ),
+        sa.PrimaryKeyConstraint(
+            "advisor_id", "student_id", name=op.f("pk_students_advisors")
+        ),
+    )
+    op.create_index(
+        "ix_students_advisors_student_id",
+        "students_advisors",
+        ["advisor_id", "student_id"],
```

Comment:
But won't the index on `student_id` and `advisor_id` be useful to this kind of search?
https://github.com/inspirehep/inspirehep/pull/2146/files/cee4526d5fd927cfc6764d5591191779980c6f74#diff-e35a8a45c8ecb58888ebe46c544224d041c296a28a295f6dc85be38d1423f916R34-R37

Reply:
I think there are several points you're misundertsanding.

* Here you're creating an index that's exactly the same as the one provided by the primary key (same columns, same order). That's never useful, it just creates a duplicate data structure. A primary key always creates an index, as the whole point of a primary key is to efficiently retrieve rows given the data in the primary key, which requires an index (and also to enforce the uniqueness constraint, which also requires an index to do efficiently).
* In a multi-column index, the ordering is important. Basically, an index (or at least a B-tree index, which is the default for non-composite fields) is a data structure that maintains a sorted list of rows. The sort order is specified by the columns. So on a `(column1, column2)` index, you can efficiently query a row given specific values for `column1` and `column2`, but also given specific values for `column1` only. However, you can't easily find rows given `column2` only. You can think of this in analogy to a dictionary (the book, not the data structure), where you need to find a definition. Because the dictionary is sorted, you can easily find a word or a prefix of a word, but finding a suffix requires scanning the whole dictionary. 
In your linked query, you're specifying `student_id` and want to get all `advisor_id`s that match, but that's exactly like trying to find a suffix given the ordering you've specified here. So you need either to define an index in the opposite order, or probably better, one containing only `student_id` (getting the rows requires going back to the main table then, but that's needed in any case because your index doesn't `degree_type`, and the index will be smaller so more efficient).

If you have some time to spare and want to learn more, I highly recommend reading https://use-the-index-luke.com/.